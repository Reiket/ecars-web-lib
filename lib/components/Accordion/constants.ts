export const ACCORDION_TEST_ID = 'accordionTestId';
