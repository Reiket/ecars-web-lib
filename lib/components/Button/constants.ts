import type {ButtonHTMLAttributes} from 'react';

export const BUTTON_SIZE = Object.freeze({
  BIG: 'big',
  SMALL: 'small',
});

export const BUTTON_COLOR = Object.freeze({
  GREEN: 'green',
  WHITE: 'white',
  GRAY: 'gray',
});


export type ButtonAttributesType = ButtonHTMLAttributes<HTMLButtonElement>['type']
export type ButtonSizeType = (typeof BUTTON_SIZE)[keyof typeof BUTTON_SIZE];
export type ButtonColorType = (typeof BUTTON_COLOR)[keyof typeof BUTTON_COLOR];

export const BUTTON_TEST_ID = 'buttonTestId';