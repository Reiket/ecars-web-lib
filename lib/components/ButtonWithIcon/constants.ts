export const LEFT_ICON_BUTTON_TEST_ID = 'leftIconButtonTestId';
export const RIGHT_ICON_BUTTON_TEST_ID = 'rightIconButtonTestId';
