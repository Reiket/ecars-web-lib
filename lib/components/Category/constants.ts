export const CATEGORIES_LIST = Object.freeze({
  GUIDES: 'guides',
  REVIEWS: 'reviews',
  NEWS: 'news',
});

export type CategoriesListType = (typeof CATEGORIES_LIST)[keyof typeof CATEGORIES_LIST];

export const CATEGORY_TEST_ID = 'categoryTestId';
