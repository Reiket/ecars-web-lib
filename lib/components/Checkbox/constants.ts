import type {ElementProps} from '@/services/types';

export interface CheckboxProps extends ElementProps {
  name?: string;
  id?: string;
  disabled?: boolean;
  onChange?: (checked: boolean) => void;
  checked?: boolean;
  hasError?: boolean;
}

export const CHECKBOX_TEST_ID = 'checkboxTestId';
export const CHECKBOX_TEST_INPUT_ID = 'checkboxTestInputId';
