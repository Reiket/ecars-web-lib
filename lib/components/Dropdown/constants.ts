import type {ReactNode} from 'react';

export interface DropdownOption {
  value: string;
  label?: ReactNode;
}

export interface DropdownProps {
  onSelect: (value: string) => void;
  category: string;
  options: DropdownOption[];
}

export const DROPDOWN_TEST_ID = 'dropdownTestId';

export const DROPDOWN_THEME = Object.freeze({
  DARK: 'dark',
  LIGHT: 'light',
});

export type DropdownThemeType = (typeof DROPDOWN_THEME)[keyof typeof DROPDOWN_THEME];
