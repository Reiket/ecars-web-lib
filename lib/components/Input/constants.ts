export const INPUT_TEST_ID = 'inputTestId';
