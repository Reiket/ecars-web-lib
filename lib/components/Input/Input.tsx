import {cn} from '@/services/helpers';
import {ChangeEvent, FC, HTMLInputTypeAttribute} from 'react';
import {INPUT_TEST_ID} from '@/components/Input/constants';
import {ElementProps} from '@/services/types';

interface Props extends ElementProps {
  type?: HTMLInputTypeAttribute;
  value?: string;
  onChange?: (value: string) => void;
  name?: string;
  placeholder?: string;
  id?: string;
  disabled?: boolean;
  hasError?: boolean;
}

export const Input: FC<Props> = ({
  value,
  onChange,
  hasError,
  placeholder,
  block,
  disabled = false,
  type = 'text',
  id = 'input',
  name = 'input',
}) => {
  const classNames = cn(block, 'input', {
    'error': hasError,
  });
  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    if (onChange) {
      onChange(event.currentTarget.value);
    }
  };
  return (
    <input
      onChange={handleChange}
      disabled={disabled}
      autoComplete="off"
      data-testid={INPUT_TEST_ID}
      className={classNames}
      id={id}
      name={name}
      placeholder={placeholder}
      value={value}
      type={type}
    />
  );
};
