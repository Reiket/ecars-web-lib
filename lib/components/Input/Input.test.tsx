import {fireEvent, render, screen} from '@testing-library/react';
import {Input} from '@/components/Input/Input';
import '@testing-library/jest-dom';
import {INPUT_TEST_ID} from '@/components/Input/constants';

describe('Input component', () => {
  const disabledOptions = [true, false];
  const errorsOptions = [true, false];
  const placeholderOptions = ['Your email', undefined];
  disabledOptions.forEach((disabled) => {
    errorsOptions.forEach((error) => {
      placeholderOptions.forEach((placeholder) => {
        const disabledStatus = disabled ? 'is disabled' : 'is not disabled';
        const hasError = error ? 'has an error and input has an error class' : 'has no error';
        const hasPlaceholder = placeholder ? 'has a placeholder' : 'has no placeholder';
        const testName = `should input ${disabledStatus} ${hasError} ${hasPlaceholder}`;
        test(testName, () => {
          const props = {
            disabled: disabled,
            hasError: error,
            placeholder: placeholder,
          };
          const {container} = render(<Input {...props} />);
          const inputElement = screen.getByTestId(INPUT_TEST_ID) as HTMLInputElement;
          if (error) {
            expect(inputElement).toHaveClass('error');
          }
          if (placeholder) {
            expect(inputElement).toHaveAttribute('placeholder', placeholder);
          }
          expect(inputElement).toBeInTheDocument();
          expect(inputElement.disabled).toBe(disabled);
          expect(container).toMatchSnapshot();
        });
      });
    });
  });
  test('should input has a value which got in props', () => {
    const props = {
      value: 'input text',
    };
    render(<Input {...props} />);
    const inputElement = screen.getByTestId(INPUT_TEST_ID);
    expect(inputElement).toHaveValue(props.value);
  });
  test('should onChange event updates value correctly', () => {
    const mockOnChange = jest.fn();
    const props = {
      value: '',
      onChange: mockOnChange,
    };
    render(<Input {...props} />);
    const inputElement = screen.getByTestId(INPUT_TEST_ID);
    fireEvent.change(inputElement, {target: {value: 'test'}});
    expect(mockOnChange).toHaveBeenCalledWith('test');
  });
  test('onChange event updates input value correctly when typing', () => {
    const mockOnChange = jest.fn();
    const props = {
      value: '',
      onChange: mockOnChange,
    };
    render(<Input {...props} />);
    const inputElement = screen.getByTestId(INPUT_TEST_ID);
    fireEvent.change(inputElement, {target: {value: 't'}});
    expect(mockOnChange).toHaveBeenCalledWith('t');
    fireEvent.change(inputElement, {target: {value: 'te'}});
    expect(mockOnChange).toHaveBeenCalledWith('te');
    fireEvent.change(inputElement, {target: {value: 'tes'}});
    expect(mockOnChange).toHaveBeenCalledWith('tes');
    fireEvent.change(inputElement, {target: {value: 'test'}});
    expect(mockOnChange).toHaveBeenCalledWith('test');
  });
});
