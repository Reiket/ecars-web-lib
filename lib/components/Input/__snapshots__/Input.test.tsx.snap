// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should input is disabled has an error and input has an invalid class has a placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input error"
    data-testid="inputTestId"
    disabled=""
    id="input"
    name="input"
    placeholder="Your email"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is disabled has an error and input has an invalid class has no placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input error"
    data-testid="inputTestId"
    disabled=""
    id="input"
    name="input"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is disabled has no error has a placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input"
    data-testid="inputTestId"
    disabled=""
    id="input"
    name="input"
    placeholder="Your email"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is disabled has no error has no placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input"
    data-testid="inputTestId"
    disabled=""
    id="input"
    name="input"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is not disabled has an error and input has an invalid class has a placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input error"
    data-testid="inputTestId"
    id="input"
    name="input"
    placeholder="Your email"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is not disabled has an error and input has an invalid class has no placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input error"
    data-testid="inputTestId"
    id="input"
    name="input"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is not disabled has no error has a placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input"
    data-testid="inputTestId"
    id="input"
    name="input"
    placeholder="Your email"
    type="text"
    value=""
  />
</div>
`;

exports[`Input component should input is not disabled has no error has no placeholder 1`] = `
<div>
  <input
    autocomplete="off"
    class="input"
    data-testid="inputTestId"
    id="input"
    name="input"
    type="text"
    value=""
  />
</div>
`;
